contracts/Compass.vy

```diff
@@ -104,39 +140,63 @@ def __init__(_compass_id: bytes32, _event_id: uint256, _gravity_nonce:uint256, v
     self.last_gravity_nonce = _gravity_nonce
     log ValsetUpdated(new_checkpoint, valset.valset_id, _event_id)
 
-# utility function to verify EIP712 signature
-@internal
-@pure
-def verify_signature(signer: address, hash: bytes32, sig: Signature) -> bool:
-    message_digest: bytes32 = keccak256(concat(convert("\x19Ethereum Signed Message:\n32", Bytes[28]), hash))
-    return signer == ecrecover(message_digest, sig.v, sig.r, sig.s)
-
-# consensus: validator set and signatures
-# hash: what we are checking they have signed
-@internal
-def check_validator_signatures(consensus: Consensus, hash: bytes32):
-    i: uint256 = 0
-    cumulative_power: uint256 = 0
-    for sig in consensus.signatures:
-        if sig.v != 0:
-            assert self.verify_signature(consensus.valset.validators[i], hash, sig), "Invalid Signature"
-            cumulative_power += consensus.valset.powers[i]
-            if cumulative_power >= POWER_THRESHOLD:
-                break
-        i = unsafe_add(i, 1)
-    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+# sends the current internal book keeping state to a new version of compass, along with all funds
+# left on the contract.
+# consensus: current validator set and signatures
+# new_compass: address of new compass to which to send the funds
+# message_id: unused message ID
+# deadline: message deadline
+# authority: intended message sender address
+# gas_estimate: gas estimation in wei
+@external
+@nonreentrant('lock')
+def send_books(consensus: Consensus, new_compass: address, message_id: uint256,deadline: uint256, authority: address, gas_estimate: uint256):
+    self.assert_authority(authority, gas_estimate)
+    assert not self.message_id_used[message_id], "Used Message_ID"
+    self.message_id_used[message_id] = True
+    # check if the supplied current validator set matches the saved checkpoint
+    assert self.last_checkpoint == self.make_checkpoint(consensus.valset), "Incorrect Checkpoint"
+    # signing data is keccak256 hash of abi_encoded logic_call(args, message_id, compass_id, deadline)
+    args_hash: bytes32 = keccak256(_abi_encode(new_compass, message_id, deadline, authority, gas_estimate, method_id=method_id("upgrade(address,uint256,uint256,address,uint256)")))
+    # check if enough validators signed args_hash
+    self.check_validator_signatures(consensus, args_hash)
+    # last fee reservation
+    self.reserve_security_refund(authority, gas_estimate)
+    self.assert_balance()
+    # Make call 
+    success, response = raw_call(
+        new_compass,
+        _abi_encode(self.funds, self.claimable_rewards, self.total_funds, self.total_claims, self.rewards_security_wallet, self.rewards_community_wallet, method_id=method_id("receive_books(HashMap[bytes32,uint256],HashMap[address,uint256],uint256,uint256,uint256,uint256)")),
```

Comment:
HashMap can't be used for a parameter of a smart contract function. That means we can't send an entire hashmap data at once. Let me think how we can implement this idea.
@byte-bandit @maharifu

Reply:
That's a problem. I didn't know that. Take the time you need to come up with a good solution, I'm not blocked.